examples/cpp/PineAPPL.hpp

```diff
@@ -0,0 +1,257 @@
+struct Order {
+    /** @brief Exponent of the strong coupling. */
+    uint32_t alphas;
+
+    /** @brief Exponent of the electromagnetic coupling. */
+    uint32_t alpha;
+
+    /** @brief Exponent of the logarithm of the scale factor of the renomalization scale. */
+    uint32_t logxir;
+
+    /** @brief Exponent of the logarithm of the scale factor of the factorization scale. */
+    uint32_t logxif;
+};
+
+class Grid {
+    /** @brief number of orders */
+    size_t n_orders = 0;
+
+    /** @brief number of bins */
+    size_t n_bins = 0;
+
+    /** @brief number of lumis */
+    size_t n_lumis = 0;
+
+public:
+
+    /** @brief underlying raw object */
+    pineappl_grid *raw;
+
+    /**
+     * @brief constructor
+     * @param lumi luminosity
+     * @param orders orders
+     * @param bins bins
+     * @param key_val additional informations
+     */
+    Grid(const Lumi& lumi, const std::vector<Order>& orders, const std::vector<double>& bins, const KeyVal& key_val) {
+        this->n_lumis = lumi.count();
+        // cast orders
+        uint32_t *raw_orders = new uint32_t[4 * orders.size()];
+        for (auto it = orders.cbegin(); it != orders.cend(); ++it, ++this->n_orders){
+            raw_orders[4*this->n_orders + 0] = it->alphas;
+            raw_orders[4*this->n_orders + 1] = it->alpha;
+            raw_orders[4*this->n_orders + 2] = it->logxir;
+            raw_orders[4*this->n_orders + 3] = it->logxif;
+        }
+        //this->n_bins = bins.size();
+        // cast bins
+        double *raw_bins = new double[bins.size()];
+        for (auto it = bins.cbegin(); it != bins.cend(); ++it, ++this->n_bins)
+            raw_bins[this->n_bins] = *it;
+        // now, we got all we need
+        this->raw = pineappl_grid_new(lumi.raw, this->n_orders, raw_orders, this->n_bins - 1, raw_bins, key_val.raw);
+        // clean up
+        delete[] raw_orders;
+        delete[] raw_bins;
+    }
+
+    /** @brief destructor */
+    ~Grid(){ pineappl_grid_delete(this->raw); }
+
+    /**
+     * @brief Fill grid for the given parameters
+     * @param x1 first momentum fraction
+     * @param x2 second momentum fraction
+     * @param q2 scale
+     * @param order order index
+     * @param observable observable value
+     * @param lumi luminosity index
+     * @param weight weight
+     */
+    void fill(double x1, double x2, double q2, size_t order, double observable, size_t lumi, double weight) const {
+        pineappl_grid_fill(this->raw, x1, x2, q2, order, observable, lumi, weight);
+    }
+
+    /**
+     * @brief perform a convolution of the grid with PDFs
+     * @param pdg_id particle ID
+     * @param pdf PDF
+     * @param xi_ren renormalization scale variation
+     * @param xi_fac factorization scale variation
+     * @return prediction for each bin
+     * @return prediction for each bin
+     */
+    std::vector<double> convolute_with_one(int32_t pdg_id, LHAPDF::PDF *pdf, double xi_ren = 1.0, double xi_fac = 1.0 ) const {
```

Comment:
Remove this member function ...

Reply:
I would prefer to keep this, because I think it matches a common use case: the variation of scales is much more common than shutting orders or lumis down (I would say). If I'm giving the possibility to overload, let's use it! I'd see it as the cpp approach to keyword arguments (in some sense). However, if you insist, I can drop it ...

Reply:
I did as you said - but now I dislike that C and C++ do not share the very same structure (which could possibly confusing)

Reply:
That's true, but at least this is something caught at compile time.